jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java

```diff
@@ -0,0 +1,205 @@
+ * the distances between each pair of cities, what is the shortest possible route that visits each
+ * city exactly once and returns to the origin city?".
+ *
+ * <p>
+ * This is an implementation of the Held-Karp algorithm which returns a optimal, minimum-cost Hamiltonian tour.
+ * The implementation requires the input graph to contain at least one vertex.
+ * The running time is $O(2^{|V|} \times |V|^2)$ and it takes $O(2^{|V|} \times |V|)$ extra memory.
+ *
+ * <p>
+ * See <a href="https://en.wikipedia.org/wiki/Travelling_salesman_problem">wikipedia</a> for more
+ * details about TSP.
+ *
+ * <p>
+ * See <a href="https://en.wikipedia.org/wiki/Held%E2%80%93Karp_algorithm">wikipedia</a> for more
+ * details about the dynamic programming algorithm.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Alexandru Valeanu
+ */
+public class HeldKarpTSP<V, E>
+        implements TSPAlgorithm<V, E> {
+
+    /**
+     * Construct a new instance
+     */
+    public HeldKarpTSP() {
+    }
+
+    private double memo(int previousNode, int state, double[][] C, double[][] W){
+        // have we seen this state before?
+        if (C[previousNode][state] != Double.MIN_VALUE)
+            return C[previousNode][state];
+
+        // no cycle has been found yet
+        double totalCost = Double.MAX_VALUE;
+
+        // check if all nodes have been visited (i.e. state + 1 == 2^n)
+        if (state == (1 << W.length) - 1){
+            // check if there is a return edge we can use
+            if (W[previousNode][0] != Double.MAX_VALUE)
+                totalCost = W[previousNode][0];
+            else
+                totalCost = Double.MAX_VALUE;
+        }
+        else{
+            // try to find the 'best' next (i.e. unvisited and adjacent to previousNode) node in the tour
+            for (int i = 0; i < W.length; i++) {
+                if (((state >> i) & 1) == 0 && W[previousNode][i] != Double.MAX_VALUE){
+                    totalCost = Math.min(totalCost, W[previousNode][i] + memo(i, state ^ (1 << i), C, W));
+                }
+            }
+        }
+
+        return C[previousNode][state] = totalCost;
+    }
+
+    /**
+     * Computes a minimum-cost Hamiltonian tour.
+     *
+     * @param graph the input graph
+     * @return a minimum-cost tour if one exists, null otherwise
+     * @throws IllegalArgumentException if the graph contains no vertices
+     * @throws IllegalArgumentException if the graph contains more than 31 vertices
+     */
+    @Override
+    public GraphPath<V, E> getTour(Graph<V, E> graph) {
+        final int n = graph.vertexSet().size(); // number of nodes
+
+        if (n == 0) {
+            throw new IllegalArgumentException("Graph contains no vertices");
+        }
+
+        if (n > 31){
+            throw new IllegalArgumentException("\"The internal representation of the dynamic programming state " +
```

Comment:
Could you drop the quote here? Not really needed.

Reply:
Dropped.